File: backend/app/schemas/personal.py
from datetime import date
from decimal import Decimal
from typing import Self
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class MomentCreate(BaseModel):
    title: str = Field(max_length=255)
    moment_type: str = Field(max_length=50)
    duration_type: str = Field(max_length=20)
    target_amount: Decimal | None = None
    start_date: date | None = None
    end_date: date | None = None
    status: str = Field(default="active", max_length=20)


class MomentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    moment_id: UUID
    user_id: str
    title: str
    moment_type: str
    duration_type: str
    target_amount: Decimal | None
    start_date: date | None
    end_date: date | None
    status: str
    created_at: str | None = None


class CycleCreate(BaseModel):
    moment_id: UUID
    label: str = Field(max_length=50)
    start_date: date
    end_date: date
    allocated_budget: Decimal = Field(default=Decimal("0"))


class CycleOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    cycle_id: UUID
    moment_id: UUID
    label: str
    start_date: date
    end_date: date
    allocated_budget: Decimal
    spent_amount: Decimal


class TransactionSubcategoryRefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    subcategory_id: UUID
    slug: str
    label: str
    sort_order: int


class TransactionCategoryTreeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    category_id: UUID
    slug: str
    label: str
    sort_order: int
    subcategories: list[TransactionSubcategoryRefOut] = Field(default_factory=list)


class TransactionCreate(BaseModel):
    amount: Decimal
    category: str | None = Field(default=None, max_length=50)
    subcategory: str | None = Field(default=None, max_length=80)
    category_id: UUID | None = None
    subcategory_id: UUID | None = None
    merchant: str | None = Field(default=None, max_length=255)
    description: str | None = None
    transaction_date: date
    moment_id: UUID | None = None
    cycle_id: UUID | None = None


class TransactionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    transaction_id: UUID
    user_id: str
    moment_id: UUID | None
    cycle_id: UUID | None
    amount: Decimal
    category: str | None
    subcategory: str | None = None
    category_id: UUID | None = None
    subcategory_id: UUID | None = None
    merchant: str | None
    description: str | None
    transaction_date: date
    created_at: str | None = None


class TransactionUpdate(BaseModel):
    amount: Decimal | None = None
    category: str | None = Field(default=None, max_length=50)
    subcategory: str | None = Field(default=None, max_length=80)
    category_id: UUID | None = None
    subcategory_id: UUID | None = None
    merchant: str | None = Field(default=None, max_length=255)
    description: str | None = None
    transaction_date: date | None = None
    moment_id: UUID | None = None
    cycle_id: UUID | None = None


class SpendBreakdownRowOut(BaseModel):
    label: str
    category_id: UUID | None = None
    amount: Decimal


class SpendBreakdownOut(BaseModel):
    rows: list[SpendBreakdownRowOut]
    total: Decimal


class BudgetCreate(BaseModel):
    cycle_id: UUID
    category: str | None = Field(default=None, max_length=50)
    category_id: UUID | None = None
    subcategory_id: UUID | None = None
    allocated_amount: Decimal

    @model_validator(mode="after")
    def require_category_scope(self) -> Self:
        has_text = bool(self.category and self.category.strip())
        if self.subcategory_id is None and self.category_id is None and not has_text:
            raise ValueError("Provide category label and/or category_id / subcategory_id")
        return self


class BudgetOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    budget_id: UUID
    cycle_id: UUID
    category: str
    subcategory: str | None = None
    allocated_amount: Decimal
    spent_amount: Decimal
    category_id: UUID | None = None
    subcategory_id: UUID | None = None


class GoalCreate(BaseModel):
    title: str = Field(max_length=255)
    target_amount: Decimal
    saved_amount: Decimal = Field(default=Decimal("0"))
    target_date: date | None = None


class GoalOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    goal_id: UUID
    user_id: str
    title: str
    target_amount: Decimal
    saved_amount: Decimal
    target_date: date | None


class SignalCreate(BaseModel):
    signal_type: str = Field(max_length=50)
    severity: str = Field(max_length=10)
    message: str


class SignalOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    signal_id: UUID
    user_id: str
    signal_type: str
    severity: str
    message: str
    created_at: str | None = None


class PersonalSummaryOut(BaseModel):
    money_left: Decimal
    total_allocated: Decimal
    total_spent_period: Decimal
    period_label: str
    insights: list[str]
    top_category: str | None
    recent_signals: list[SignalOut]

File: frontend/lib/api/personal.ts
import { getApiBaseUrl } from "@/lib/api/client";

async function headersJson(token: string) {
  return {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
  } as const;
}

async function parseJson<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const text = await res.text().catch(() => "");
    if (text) {
      try {
        const json = JSON.parse(text) as Record<string, unknown>;
        if (typeof json.detail === "string") throw new Error(json.detail);
        if (typeof json.message === "string") throw new Error(json.message);
      } catch (e) {
        if (e instanceof Error && e.message !== text) throw e;
      }
    }
    throw new Error(text || `${res.status}`);
  }
  return res.json() as Promise<T>;
}

export type PersonalSummary = {
  money_left: string | number;
  total_allocated: string | number;
  total_spent_period: string | number;
  period_label: string;
  insights: string[];
  top_category: string | null;
  recent_signals: {
    signal_id: string;
    user_id: string;
    signal_type: string;
    severity: string;
    message: string;
    created_at?: string | null;
  }[];
};

export type PersonalMoment = {
  moment_id: string;
  user_id: string;
  title: string;
  moment_type: string;
  duration_type: string;
  target_amount: number | null;
  start_date: string | null;
  end_date: string | null;
  status: string;
  created_at?: string | null;
};

export type PersonalCycle = {
  cycle_id: string;
  moment_id: string;
  label: string;
  start_date: string;
  end_date: string;
  allocated_budget: number;
  spent_amount: number;
};

export type PersonalGoal = {
  goal_id: string;
  user_id: string;
  title: string;
  target_amount: number;
  saved_amount: number;
  target_date: string | null;
};

export type PersonalBudget = {
  budget_id: string;
  cycle_id: string;
  category: string;
  subcategory?: string | null;
  allocated_amount: number;
  spent_amount: number;
  category_id?: string | null;
  subcategory_id?: string | null;
};

export type PersonalTxnSubcategory = {
  subcategory_id: string;
  slug: string;
  label: string;
  sort_order: number;
};

export type PersonalTxnCategory = {
  category_id: string;
  slug: string;
  label: string;
  sort_order: number;
  subcategories: PersonalTxnSubcategory[];
};

export type PersonalTransaction = {
  transaction_id: string;
  user_id: string;
  moment_id: string | null;
  cycle_id: string | null;
  amount: number;
  category: string | null;
  subcategory?: string | null;
  category_id?: string | null;
  subcategory_id?: string | null;
  merchant: string | null;
  description: string | null;
  transaction_date: string;
  created_at?: string | null;
};

export type SpendBreakdown = {
  rows: { label: string; category_id: string | null; amount: string | number }[];
  total: string | number;
};

export async function fetchPersonalSummary(token: string): Promise<PersonalSummary> {
  const res = await fetch(`${getApiBaseUrl()}/personal/summary`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<PersonalSummary>(res);
}

export async function fetchMoments(token: string): Promise<PersonalMoment[]> {
  const res = await fetch(`${getApiBaseUrl()}/personal/moments`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<PersonalMoment[]>(res);
}

export async function fetchCycles(token: string, momentId?: string): Promise<PersonalCycle[]> {
  const q = momentId ? `?moment_id=${encodeURIComponent(momentId)}` : "";
  const res = await fetch(`${getApiBaseUrl()}/personal/cycles${q}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<PersonalCycle[]>(res);
}

export type FetchTransactionsOpts = {
  limit?: number;
  month?: string;
  cycle_id?: string;
  category_id?: string;
  merchant?: string;
};

export async function fetchTransactions(
  token: string,
  opts?: FetchTransactionsOpts,
): Promise<PersonalTransaction[]> {
  const q = new URLSearchParams();
  q.set("limit", String(opts?.limit ?? 100));
  if (opts?.month) q.set("month", opts.month);
  if (opts?.cycle_id) q.set("cycle_id", opts.cycle_id);
  if (opts?.category_id) q.set("category_id", opts.category_id);
  if (opts?.merchant?.trim()) q.set("merchant", opts.merchant.trim());
  const res = await fetch(`${getApiBaseUrl()}/personal/transactions?${q}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<PersonalTransaction[]>(res);
}

export async function fetchSpendBreakdown(
  token: string,
  opts?: { cycle_id?: string; month?: string },
): Promise<SpendBreakdown> {
  const q = new URLSearchParams();
  if (opts?.month) q.set("month", opts.month);
  if (opts?.cycle_id) q.set("cycle_id", opts.cycle_id);
  const qs = q.toString();
  const res = await fetch(
    `${getApiBaseUrl()}/personal/spend-breakdown${qs ? `?${qs}` : ""}`,
    { headers: { Authorization: `Bearer ${token}` } },
  );
  return parseJson<SpendBreakdown>(res);
}

export async function fetchTransactionCategories(
  token: string,
): Promise<PersonalTxnCategory[]> {
  const res = await fetch(`${getApiBaseUrl()}/personal/transaction-categories`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<PersonalTxnCategory[]>(res);
}

export async function createTransaction(
  token: string,
  body: {
    amount: number;
    category?: string | null;
    subcategory?: string | null;
    category_id?: string | null;
    subcategory_id?: string | null;
    merchant?: string | null;
    description?: string | null;
    transaction_date: string;
    moment_id?: string | null;
    cycle_id?: string | null;
  },
): Promise<PersonalTransaction> {
  const res = await fetch(`${getApiBaseUrl()}/personal/transactions`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<PersonalTransaction>(res);
}

export async function updateTransaction(
  token: string,
  transactionId: string,
  body: Partial<{
    amount: number;
    category: string | null;
    subcategory: string | null;
    category_id: string | null;
    subcategory_id: string | null;
    merchant: string | null;
    description: string | null;
    transaction_date: string;
    moment_id: string | null;
    cycle_id: string | null;
  }>,
): Promise<PersonalTransaction> {
  const res = await fetch(`${getApiBaseUrl()}/personal/transactions/${transactionId}`, {
    method: "PATCH",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<PersonalTransaction>(res);
}

export async function deleteTransaction(token: string, transactionId: string): Promise<void> {
  const res = await fetch(`${getApiBaseUrl()}/personal/transactions/${transactionId}`, {
    method: "DELETE",
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) {
    return parseJson<void>(res);
  }
}

export async function fetchBudgets(token: string, cycleId: string): Promise<PersonalBudget[]> {
  const q = new URLSearchParams({ cycle_id: cycleId });
  const res = await fetch(`${getApiBaseUrl()}/personal/budgets?${q}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<PersonalBudget[]>(res);
}

export async function createBudget(
  token: string,
  body: {
    cycle_id: string;
    allocated_amount: number;
    category?: string | null;
    category_id?: string | null;
    subcategory_id?: string | null;
  },
): Promise<PersonalBudget> {
  const res = await fetch(`${getApiBaseUrl()}/personal/budgets`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<PersonalBudget>(res);
}

export async function createCycle(
  token: string,
  body: {
    moment_id: string;
    label: string;
    start_date: string;
    end_date: string;
    allocated_budget: number;
  },
): Promise<PersonalCycle> {
  const res = await fetch(`${getApiBaseUrl()}/personal/cycles`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<PersonalCycle>(res);
}

export async function createMoment(
  token: string,
  body: {
    title: string;
    moment_type: string;
    duration_type: string;
    target_amount?: number | null;
    start_date?: string | null;
    end_date?: string | null;
    status?: string;
  },
): Promise<PersonalMoment> {
  const res = await fetch(`${getApiBaseUrl()}/personal/moments`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify(body),
  });
  return parseJson<PersonalMoment>(res);
}

export async function fetchGoals(token: string): Promise<PersonalGoal[]> {
  const res = await fetch(`${getApiBaseUrl()}/personal/goals`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<PersonalGoal[]>(res);
}

export async function createGoal(
  token: string,
  body: {
    title: string;
    target_amount: number;
    saved_amount?: number;
    target_date?: string | null;
  },
): Promise<PersonalGoal> {
  const res = await fetch(`${getApiBaseUrl()}/personal/goals`, {
    method: "POST",
    headers: await headersJson(token),
    body: JSON.stringify({
      title: body.title,
      target_amount: body.target_amount,
      saved_amount: body.saved_amount ?? 0,
      target_date: body.target_date ?? undefined,
    }),
  });
  return parseJson<PersonalGoal>(res);
}

export async function evaluateSignals(token: string): Promise<{ insights: number }> {
  const res = await fetch(`${getApiBaseUrl()}/personal/signals/evaluate`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}` },
  });
  return parseJson<{ insights: number }>(res);
}

export function transactionsToCsv(rows: PersonalTransaction[]): string {
  const header = [
    "transaction_id",
    "transaction_date",
    "amount",
    "category",
    "subcategory",
    "merchant",
    "description",
    "cycle_id",
  ];
  const esc = (s: string | null | undefined) => {
    const v = (s ?? "").replace(/"/g, '""');
    return `"${v}"`;
  };
  const lines = [header.join(",")];
  for (const t of rows) {
    lines.push(
      [
        esc(t.transaction_id),
        esc(t.transaction_date),
        String(t.amount),
        esc(t.category),
        esc(t.subcategory ?? ""),
        esc(t.merchant),
        esc(t.description),
        esc(t.cycle_id),
      ].join(","),
    );
  }
  return lines.join("\n");
}
